tests/unit-tests/wayland/test_wayland_weak.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright © 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include "mir/wayland/wayland_base.h"
+
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+#include <experimental/optional>
+
+namespace mw = mir::wayland;
+
+using namespace testing;
+
+class MockResource : public virtual mw::LifetimeTracker
+{
+public:
+    // Expose protected method
+    void mark_destroyed() const
+    {
+        LifetimeTracker::mark_destroyed();
+    }
+};
+
+template<typename T>
+auto operator<<(std::ostream& out, mw::Weak<T> const& weak) -> std::ostream&
+{
+    return out << "wayland::Weak{" << (weak ? &weak.value() : nullptr) << "}";
+}
+
+class WaylandWeakTest : public Test
+{
+public:
+    std::experimental::optional<MockResource> resource{{}};
+    std::experimental::optional<MockResource> resource_a{{}};
+    std::experimental::optional<MockResource> resource_b{{}};
+};
+
+TEST_F(WaylandWeakTest, returns_resource)
+{
+    mw::Weak<MockResource> const weak{&resource.value()};
+    ASSERT_THAT(&weak.value(), Eq(&resource.value()));
+}
+
+TEST_F(WaylandWeakTest, false_after_resource_destroyed)
+{
+    mw::Weak<MockResource> const weak{&resource.value()};
+    ASSERT_THAT(weak, Eq(true));
+    resource = std::experimental::nullopt;
+    ASSERT_THAT(weak, Eq(false));
+}
+
+TEST_F(WaylandWeakTest, can_be_created_with_nullptr)
+{
+    mw::Weak<MockResource> const weak{nullptr};
+    ASSERT_THAT(weak, Eq(false));
+}
+
+TEST_F(WaylandWeakTest, throws_logic_error_on_access_after_resource_destroyed)
+{
+    mw::Weak<MockResource> const weak{&resource.value()};
+    resource = std::experimental::nullopt;
+    EXPECT_THROW(weak.value(), std::logic_error);
+}
+
+TEST_F(WaylandWeakTest, throws_logic_error_on_access_of_null)
+{
+    mw::Weak<MockResource> const weak;
+    EXPECT_THROW(weak.value(), std::logic_error);
+}
+
+TEST_F(WaylandWeakTest, copies_are_equal)
+{
+    mw::Weak<MockResource> const weak_a{&resource.value()};
+    mw::Weak<MockResource> const weak_b{weak_a};
+    EXPECT_THAT(weak_a, Eq(weak_b));
+    EXPECT_THAT(weak_b, Eq(weak_a));
```

Comment:
Maybe for all of these we should be testing the full matrix? ie:
a == b : True
b == a : True
a != b: False
b != a : False
?